ext/nnstreamer/extra/nnstreamer_grpc_protobuf.cc

```diff
@@ -0,0 +1,224 @@
+}
+
+template <typename T>
+Status ServiceImplProtobuf::_write_tensors (T writer)
+{
+  GstDataQueueItem *item;
+
+  /* until flushing */
+  while (gst_data_queue_pop (queue_, &item)) {
+    Tensors tensors;
+
+    _get_tensors_from_buffer (GST_BUFFER (item->object), tensors);
+    writer->Write (tensors);
+
+    g_free (item);
+  }
+
+  return Status::OK;
+}
+
+template <typename T>
+Status ServiceImplProtobuf::_read_tensors (T reader)
+{
+  Tensors tensors;
+
+  while (reader->Read (&tensors)) {
+    GstBuffer *buffer;
+
+    _get_buffer_from_tensors (tensors, &buffer);
+
+    if (cb_)
+      cb_ (cb_data_, buffer);
+  }
+
+  return Status::OK;
+}
+
+/** @brief gRPC client thread */
+void
+ServiceImplProtobuf::_client_thread ()
+{
+  ClientContext context;
+  Empty empty;
+
+  if (direction_ == GRPC_DIRECTION_TENSORS_TO_BUFFER) {
+    /* initiate the RPC call */
+    std::unique_ptr< ClientWriter<Tensors> > writer(
+        client_stub_->SendTensors (&context, &empty));
+
+    _write_tensors (writer.get ());
+
+    writer->WritesDone ();
+    writer->Finish ();
+  } else if (direction_ == GRPC_DIRECTION_BUFFER_TO_TENSORS) {
+    Tensors tensors;
+
+    /* initiate the RPC call */
+    std::unique_ptr< ClientReader<Tensors> > reader(
+        client_stub_->RecvTensors (&context, empty));
+
+    _read_tensors (reader.get ());
+
+    reader->Finish ();
+  } else {
+    g_assert (0); /* internal logic error */
+  }
+}
+
+/** @brief start gRPC server handling protobuf */
+gboolean
+ServiceImplProtobuf::start_server (std::string address)
+{
+  /* listen on the given address without any authentication mechanism */
+  ServerBuilder builder;
+  builder.AddListeningPort (address, grpc::InsecureServerCredentials(), &port_);
+  builder.RegisterService (this);
+
+  /* start the server */
+  server_worker_ = builder.BuildAndStart ();
+  if (server_worker_.get () == nullptr)
+    return FALSE;
+
+  return TRUE;
+}
+
+/** @brief start gRPC client handling protobuf */
+gboolean
+ServiceImplProtobuf::start_client (std::string address)
+{
+  /* create a gRPC channel */
+  std::shared_ptr<Channel> channel = grpc::CreateChannel(
+      address, grpc::InsecureChannelCredentials());
+
+  /* connect the server */
+  client_stub_ = TensorService::NewStub (channel);
+  if (client_stub_.get () == nullptr)
+    return FALSE;
+
+  client_worker_ = std::thread ([this] { this->_client_thread (); });
+
+  return TRUE;
+}
+
+/** @brief convert tensors to buffer */
+void
+ServiceImplProtobuf::_get_buffer_from_tensors (Tensors &tensors,
+    GstBuffer **buffer)
+{
+  guint num_tensor = tensors.num_tensor ();
+  GstMemory *memory;
+
+  *buffer = gst_buffer_new ();
+
+  for (guint i = 0; i < num_tensor; i++) {
+    const Tensor * tensor = &tensors.tensor (i);
+    const void * data = tensor->data ().c_str ();
+    gsize size = tensor->data ().length ();
+
+    memory = gst_memory_new_wrapped (GST_MEMORY_FLAG_READONLY,
+        g_memdup (data, size), size, 0, size, NULL, NULL);
```

Comment:
this needs g_free to release mem block:
gpointer new_mem = g_memdup (data, size);
gst_memory_new_wrapped (..., new_mem, g_free)

Reply:
Oh, thanks. I missed to handle this. In the original codes, the callers had managed its de-allocation, but now it's no longer valid.